File: FPSCounter-java/src/net/sabamiso/java/utils/FPSCounter.java
//
//	FPSCounter.java - simple performance check class for fps (frame per second).
//
// github:
//     https://github.com/yoggy/FPSCounter
//
// license:
//     Copyright (c) 2015 yoggy <yoggy0@gmail.com>
//     Released under the MIT license
//     http://opensource.org/licenses/mit-license.php
//
package net.sabamiso.java.utils;

public class FPSCounter {
	String name;
	int check_count;
	boolean debug_output;

	long st;
	int count;
	float t;
	float fps;

	public FPSCounter() {
		this.name = "FPSCounter";
		this.check_count = 100;
		this.debug_output = false;

		clear();
	}

	public FPSCounter(String name) {
		this.name = name;
		this.check_count = 100;
		this.debug_output = false;

		clear();
	}

	public FPSCounter(String name, int check_count) {
		this.name = name;
		this.check_count = check_count;
		this.debug_output = false;

		clear();
	}

	public FPSCounter(String name, int check_count, boolean debug_output) {
		this.name = name;
		this.check_count = check_count;
		this.debug_output = debug_output;

		clear();
	}

	public void clear() {
		st = getTickCount();
		count = 0;
		t = 0.0f;
		fps = 0.0f;
	}

	public String getName() {
		return name;
	}

	public void setName(String val) {
		name = val;
	}

	public int getCheckCount() {
		return check_count;
	}

	public void setCheckCount(int val) {
		if (val <= 0) {
			System.out
					.println("ERROR : FPSCounter.setCheckCount(int val) : invalid val...val="
							+ val);
			return;
		}
		check_count = val;
		clear();
	}

	public boolean getDebugOutput() {
		return debug_output;
	}

	public void setDebugOutput(boolean flag) {
		debug_output = flag;
	}

	public float getT() {
		return t;
	}

	public float fps() {
		return fps;
	}

	public void check() {
		count++;
		if (count == check_count) {
			long et = getTickCount();
			long diff_tick = et - st;

			float diff_sec = diff_tick / 1000.0f;

			t = diff_sec / (float) check_count;
			fps = 1.0f / t;

			st = et;
			count = 0;

			if (debug_output) {
				outputDebugString();
			}
		}
	}

	public void outputDebugString() {
		System.out.println(toString());
	}

	public String toString() {
		String str = "" + name + " : t=" + t + ", fps=" + fps;
		return str;
	}

	protected long getTickCount() {
		return System.currentTimeMillis();
	}
	
	//
	// How to use...
	//
	public static void main(String [] args) {
		FPSCounter fps = new FPSCounter();
		fps.setDebugOutput(true);
		
		while(true) {
			try {
				Thread.sleep(33);
			} catch (InterruptedException e) {
			}
			
			fps.check();
		}
	}
}

File: FPSCounter-boost/FPSCounter.cpp
//
//	FPSCounter.{h|cpp} - simple performance check class for fps (frame per second).
//
// license:
// Copyright (c) 2015 yoggy <yoggy0@gmail.com>
// Released under the MIT license
// http://opensource.org/licenses/mit-license.php
//
#include "FPSCounter.h"

#include <iostream>
#include <sstream>
#include <limits>

FPSCounter::FPSCounter(const std::string &name, const int &check_count, const bool &debug_output)
	: name_(name), check_count_(check_count), debug_output_(debug_output)
{
	clear();
}

FPSCounter::FPSCounter(const FPSCounter &obj)
{
	copy(obj, *this);
}

FPSCounter& FPSCounter::operator=(const FPSCounter &obj)
{
	copy(obj, *this);
	return *this;
}

void FPSCounter::copy(const FPSCounter &src, FPSCounter &dst)
{
	dst.name_ = src.name_;
	dst.check_count_ = src.check_count_;
	dst.debug_output_ = src.debug_output_;

	dst.st_ = src.st_;
	dst.count_ = src.count_;
	dst.t_ = src.t_;
	dst.fps_ = src.fps_;
}

FPSCounter::~FPSCounter()
{

}

void FPSCounter::clear()
{
	st_ = get_tick_count_();
	count_ = 0;
	t_ = 0;
	fps_ = 0;
}

std::string FPSCounter::name() const
{
	return name_;
}

void FPSCounter::name(const std::string &val)
{
	name_ = val;
}

int FPSCounter::check_count() const
{
	return check_count_;
}

void FPSCounter::check_count(const int &val)
{
	if (val <= 0) {
		printf("ERROR : FPSCounter::check_count(const int &val) : invalid val...val=%d\n", val);
		return;
	}
	check_count_ = val;
	clear();
}

bool FPSCounter::debug_output() const
{
	return debug_output_;
}

void FPSCounter::debug_output(const bool &flag)
{
	debug_output_ = flag;
}

float FPSCounter::t() const
{
	return t_;
}

float FPSCounter::fps() const
{
	return fps_;
}

void FPSCounter::check()
{
	count_++;
	if (count_ == check_count_) {
		FPS_COUNTER_TICK_TYPE et = get_tick_count_();
		FPS_COUNTER_TICK_TYPE diff_tick = et - st_;

		float diff_sec = diff_tick / (float)FPS_COUNTER_TICK_PRECISION;

		t_ = diff_sec / (float)check_count_;
		fps_ = 1.0f / t_;

		st_ = et;
		count_ = 0;

		if (debug_output_) {
			output_status_string();
		}
	}
}

void FPSCounter::output_status_string() const
{
	std::cout << status_string() << std::endl;
}

std::string FPSCounter::status_string() const
{
	std::stringstream ss;
	ss.precision(4);
	ss << name() << " : " << "t=" << t() << ", fps=" << fps();

	return ss.str();
}

boost::posix_time::ptime epoch_19700101(boost::gregorian::date(1970, 1, 1));

FPS_COUNTER_TICK_TYPE FPSCounter::get_tick_count_()
{
	boost::posix_time::ptime t = boost::posix_time::microsec_clock::local_time();
	boost::posix_time::time_duration d = t - epoch_19700101;
	return d.total_milliseconds(); 
}

